### PR TITLE
[SPARK-18705][ML][DOC] Update user guide to reflect one pass solver for L1 and elastic-net

### PR DESCRIPTION
**Pull request by sethah**
## What changes were proposed in this pull request?

WeightedLeastSquares now supports L1 and elastic net penalties and has an additional solver option: QuasiNewton. The docs are updated to reflect this change.

## How was this patch tested?

Docs only. Generated documentation to make sure Latex looks ok.

### CODE REVIEW

File: docs/ml-advanced.md [hunk 1]
```diff
@@ -59,17 +59,22 @@ Given $n$ weighted observations $(w_i, a_i, b_i)$:
...
+\min_{\mathbf{x}}\frac{1}{2} \sum_{i=1}^n \frac{w_i(\mathbf{a}_i^T \mathbf{x} -b_i)^2}{\sum_{k=1}^n w_k} + \frac{1}{2}\frac{\lambda}{\delta}\sum_{j=1}^m(\sigma_{j} x_{j})^2
 \]`
 where $\lambda$ is the regularization parameter, $\delta$ is the population standard deviation of the label
 and $\sigma_j$ is the population standard deviation of the j-th feature column.
 
-This objective function has an analytic solution and it requires only one pass over the data to collect necessary statistics to solve.
-Unlike the original dataset which can only be stored in a distributed system,
-these statistics can be loaded into memory on a single machine if the number of features is relatively small, and then we can solve the objective function through Cholesky factorization on the driver.
+This objective function has an analytic solution and it requires only one pass over the data to collect necessary statistics to solve. For an
+$n \times m$ data matrix, these statistics require only $O(m^2)$ storage and so can be stored on a single machine when $n$ (the number of features) is
```
Comment (MLnick) [inline docs/ml-advanced.md]: `$n$ (the number of features)` -> `$m$ (the number of features)`?

Comment (sethah) [inline docs/ml-advanced.md]: Yep, thanks!

**Comment by sethah** [comment]
ping @yanboliang

File: docs/ml-advanced.md [hunk 2]
```diff
@@ -59,17 +59,22 @@ Given $n$ weighted observations $(w_i, a_i, b_i)$:
 
 The number of features for each observation is $m$. We use the following weighted least squares formulation:
 `\[   
-minimize_{x}\frac{1}{2} \sum_{i=1}^n \frac{w_i(a_i^T x -b_i)^2}{\sum_{k=1}^n w_k} + \frac{1}{2}\frac{\lambda}{\delta}\sum_{j=1}^m(\sigma_{j} x_{j})^2
+\min_{\mathbf{x}}\frac{1}{2} \sum_{i=1}^n \frac{w_i(\mathbf{a}_i^T \mathbf{x} -b_i)^2}{\sum_{k=1}^n w_k} + \frac{1}{2}\frac{\lambda}{\delta}\sum_{j=1}^m(\sigma_{j} x_{j})^2
```
Comment (yanboliang) [inline docs/ml-advanced.md]: This formulation is out of date, we should update it to include L1/elasticNet regularization.

File: docs/ml-advanced.md [hunk 3]
```diff
@@ -59,17 +59,22 @@ Given $n$ weighted observations $(w_i, a_i, b_i)$:
...
-these statistics can be loaded into memory on a single machine if the number of features is relatively small, and then we can solve the objective function through Cholesky factorization on the driver.
+This objective function has an analytic solution and it requires only one pass over the data to collect necessary statistics to solve. For an
+$n \times m$ data matrix, these statistics require only $O(m^2)$ storage and so can be stored on a single machine when $m$ (the number of features) is
+relatively small. We can then solve the normal equations on a single machine using local methods like direct Cholesky factorization or iterative optimization programs.
 
-WeightedLeastSquares only supports L2 regularization and provides options to enable or disable regularization and standardization.
-In order to make the normal equation approach efficient, WeightedLeastSquares requires that the number of features be no more than 4096. For larger problems, use L-BFGS instead.
+Spark ML currently supports two types of solvers for the normal equations: Cholesky factorization and Quasi-Newton methods (L-BFGS/OWL-QN). Cholesky factorization
```
Comment (yanboliang) [inline docs/ml-advanced.md]: ```ML``` -> ```MLlib```, "Spark ML" is not an official name of the component.

File: docs/ml-advanced.md [hunk 4]
```diff
@@ -59,17 +59,22 @@ Given $n$ weighted observations $(w_i, a_i, b_i)$:
...
+depends on a positive definite covariance matrix (e.g. columns of the data matrix must be linearly independent) and will fail if this condition is violated. Quasi-Newton methods
+are still capable of providing a reasonable solution even when the covariance matrix is not positive definite, so the normal equation solver can also fall back to 
+Quasi-Newton methods in this case. This fallback is currently always enabled for the `LinearRegression` estimator.
```
Comment (yanboliang) [inline docs/ml-advanced.md]: and ```GeneralizedLinearRegression``` estimator.

File: docs/ml-advanced.md [hunk 5]
```diff
@@ -59,17 +59,22 @@ Given $n$ weighted observations $(w_i, a_i, b_i)$:
...
+
+`WeightedLeastSquares` supports L1, L2, and elastic-net regularization and provides options to enable or disable regularization and standardization.
```
Comment (yanboliang) [inline docs/ml-advanced.md]: Adding following clarification should be more clear?
* For L2 or no regularization, Cholesky solver is the default choice and will fall back to Quasi-Newton solver if the covariance matrix is not positive definite.
* For L1/elasticNet regularization, Quasi-Newton solver is the default and only choice.

**Review by yanboliang — COMMENTED** [review]
@sethah Thanks for working on this, I left some comments.

File: docs/ml-advanced.md [hunk 6]
```diff
@@ -59,17 +59,25 @@ Given $n$ weighted observations $(w_i, a_i, b_i)$:
...
+This objective function requires only one pass over the data to collect the statistics necessary to solve it. For an
+$n \times m$ data matrix, these statistics require only $O(m^2)$ storage and so can be stored on a single machine when $m$ (the number of features) is
+relatively small. We can then solve the normal equations on a single machine using local methods like direct Cholesky factorization or iterative optimization programs.
 
-WeightedLeastSquares only supports L2 regularization and provides options to enable or disable regularization and standardization.
-In order to make the normal equation approach efficient, WeightedLeastSquares requires that the number of features be no more than 4096. For larger problems, use L-BFGS instead.
+Spark MLlib currently supports two types of solvers for the normal equations: Cholesky factorization and Quasi-Newton methods (L-BFGS/OWL-QN). Cholesky factorization
+depends on a positive definite covariance matrix (e.g. columns of the data matrix must be linearly independent) and will fail if this condition is violated. Quasi-Newton methods
```
Comment (jkbradley) [inline docs/ml-advanced.md]: "e.g." -> "i.e."

**Review by jkbradley — COMMENTED** [review]
LGTM except for the one nit.

File: docs/ml-advanced.md (hunk 6 above)
Comment (sethah) [inline docs/ml-advanced.md]: Done, thanks!

File: docs/ml-advanced.md (hunk 5 above)
Comment (jkbradley) [inline docs/ml-advanced.md]: I think the current note suffices.

Comment (yanboliang) [inline docs/ml-advanced.md]: The audience here is developers/contributors, and the current change is perfectly OK.

**Comment by yanboliang** [comment]
LGTM, merged into master and branch-2.1. Thanks for all.

**Pull request closed without merging**